# Mixed Logit

In [1]:
import pandas as pd
import numpy as np
import jax

from jaxlogit.mixed_logit import MixedLogit

In [2]:
#  64bit precision
jax.config.update("jax_enable_x64", True)

## Swissmetro Dataset


The swissmetro dataset contains stated-preferences for three alternative transportation modes that include car, train and a newly introduced mode: the swissmetro. This dataset is commonly used for estimation examples with the `Biogeme` and `PyLogit` packages. The dataset is available at http://transp-or.epfl.ch/data/swissmetro.dat and [Bierlaire et. al., (2001)](https://transp-or.epfl.ch/documents/proceedings/BierAxhaAbay01.pdf) provides a detailed discussion of the data as wells as its context and collection process. The explanatory variables in this example include the travel time (`TT`) and cost `CO` for each of the three alternative modes.

### Read data

The dataset is imported to the Python environment using `pandas`. Then, two types of samples, ones with a trip purpose different to commute or business and ones with an unknown choice, are filtered out. The original dataset contains 10,729 records, but after filtering, 6,768 records remain for following analysis. Finally, a new column that uniquely identifies each sample is added to the dataframe and the `CHOICE` column, which originally contains a numerical coding of the choices, is mapped to a description that is consistent with the alternatives in the column names. 

In [3]:
df_wide = pd.read_table("http://transp-or.epfl.ch/data/swissmetro.dat", sep='\t')

# Keep only observations for commute and business purposes that contain known choices
df_wide = df_wide[(df_wide['PURPOSE'].isin([1, 3]) & (df_wide['CHOICE'] != 0))]

df_wide['custom_id'] = np.arange(len(df_wide))  # Add unique identifier
df_wide['CHOICE'] = df_wide['CHOICE'].map({1: 'TRAIN', 2:'SM', 3: 'CAR'})
df_wide

GROUP  SURVEY  SP   ID  PURPOSE  FIRST  TICKET  WHO  LUGGAGE  AGE  ...  \
0         2       0   1    1        1      0       1    1        0    3  ...   
1         2       0   1    1        1      0       1    1        0    3  ...   
2         2       0   1    1        1      0       1    1        0    3  ...   
3         2       0   1    1        1      0       1    1        0    3  ...   
4         2       0   1    1        1      0       1    1        0    3  ...   
...     ...     ...  ..  ...      ...    ...     ...  ...      ...  ...  ...   
8446      3       1   1  939        3      1       7    3        1    5  ...   
8447      3       1   1  939        3      1       7    3        1    5  ...   
8448      3       1   1  939        3      1       7    3        1    5  ...   
8449      3       1   1  939        3      1       7    3        1    5  ...   
8450      3       1   1  939        3      1       7    3        1    5  ...   

      TRAIN_CO  TRAIN_HE  SM_TT  SM_CO  SM_HE  SM_SEATS  CAR_TT  CAR_CO  \
0           48       120     63     52     20         0     117      65   
1           48        30     60     49     10         0     117      84   
2           48        60     67     58     30         0     117      52   
3           40        30     63     52     20         0      72      52   
4           36        60     63     42     20         0      90      84   
...        ...       ...    ...    ...    ...       ...     ...     ...   
8446        13        30     50     17     30         0     130      64   
8447        12        30     53     16     10         0      80      80   
8448        16        60     50     16     20         0      80      64   
8449        16        30     53     17     30         0      80     104   
8450        13        60     53     21     30         0     100      80   

      CHOICE  custom_id  
0         SM          0  
1         SM          1  
2         SM          2  
3         SM          3  
4         SM          4  
...      ...        ...  
8446   TRAIN       6763  
8447   TRAIN       6764  
8448   TRAIN       6765  
8449   TRAIN       6766  
8450   TRAIN       6767  

[6768 rows x 29 columns]

### Reshape data

The imported dataframe is in wide format, and it needs to be reshaped to long format for processing by `xlogit`, which offers the convenient `wide_to_long` utility for this reshaping process. The user needs to specify the column that uniquely identifies each sample, the names of the alternatives, the columns that vary across alternatives, and whether the alternative names are a prefix or suffix of the column names. Additionally, the user can specify a value (`empty_val`) to be used by default when an alternative is not available for a certain variable. Additional usage examples for the `wide_to_long` function are available in xlogit's documentation at https://xlogit.readthedocs.io/en/latest/notebooks/convert_data_wide_to_long.html. Also, details about the function parameters are available at the [API reference ](https://xlogit.readthedocs.io/en/latest/api/utils.html#xlogit.utils.wide_to_long).

In [4]:
from jaxlogit.utils import wide_to_long

df = wide_to_long(df_wide, id_col='custom_id', alt_name='alt', sep='_',
                  alt_list=['TRAIN', 'SM', 'CAR'], empty_val=0,
                  varying=['TT', 'CO', 'HE', 'AV', 'SEATS'], alt_is_prefix=True)
df

custom_id    alt   TT   CO   HE  AV  SEATS  GROUP  SURVEY  SP  ...  \
0              0  TRAIN  112   48  120   1      0      2       0   1  ...   
1              0     SM   63   52   20   1      0      2       0   1  ...   
2              0    CAR  117   65    0   1      0      2       0   1  ...   
3              1  TRAIN  103   48   30   1      0      2       0   1  ...   
4              1     SM   60   49   10   1      0      2       0   1  ...   
...          ...    ...  ...  ...  ...  ..    ...    ...     ...  ..  ...   
20299       6766     SM   53   17   30   1      0      3       1   1  ...   
20300       6766    CAR   80  104    0   1      0      3       1   1  ...   
20301       6767  TRAIN  108   13   60   1      0      3       1   1  ...   
20302       6767     SM   53   21   30   1      0      3       1   1  ...   
20303       6767    CAR  100   80    0   1      0      3       1   1  ...   

       TICKET  WHO  LUGGAGE  AGE  MALE  INCOME  GA  ORIGIN  DEST  CHOICE  
0           1    1        0    3     0       2   0       2     1      SM  
1           1    1        0    3     0       2   0       2     1      SM  
2           1    1        0    3     0       2   0       2     1      SM  
3           1    1        0    3     0       2   0       2     1      SM  
4           1    1        0    3     0       2   0       2     1      SM  
...       ...  ...      ...  ...   ...     ...  ..     ...   ...     ...  
20299       7    3        1    5     1       2   0       1     2   TRAIN  
20300       7    3        1    5     1       2   0       1     2   TRAIN  
20301       7    3        1    5     1       2   0       1     2   TRAIN  
20302       7    3        1    5     1       2   0       1     2   TRAIN  
20303       7    3        1    5     1       2   0       1     2   TRAIN  

[20304 rows x 23 columns]

### Create model specification

Following the reshaping, users can create or update the dataset's columns in order to accommodate their model specification needs, if necessary. The code below shows how the columns `ASC_TRAIN` and `ASC_CAR` were created to incorporate alternative-specific constants in the model. In addition, the example illustrates an effective way of establishing variable interactions (e.g., trip costs for commuters with an annual pass) by updating existing columns conditional on values of other columns. Although apparently simple, column operations provide users with an intuitive and highly-flexible mechanism to incorporate model specification aspects, such as variable transformations, interactions, and alternative specific coefficients and constants. By operating the dataframe columns, any utility specification can be accommodated in `xlogit`. As shown in [this specification example](https://xlogit.readthedocs.io/en/latest/notebooks/multinomial_model.html#Create-model-specification), highly-flexible utility specifications can be modeled in `xlogit` by operating the dataframe columns.

In [5]:
df['ASC_TRAIN'] = np.ones(len(df))*(df['alt'] == 'TRAIN')
df['ASC_CAR'] = np.ones(len(df))*(df['alt'] == 'CAR')
df['TT'], df['CO'] = df['TT']/100, df['CO']/100  # Scale variables
annual_pass = (df['GA'] == 1) & (df['alt'].isin(['TRAIN', 'SM']))
df.loc[annual_pass, 'CO'] = 0  # Cost zero for pass holders

### Estimate model parameters

The `fit` method estimates the model by taking as input the data from the previous step along with additional specification criteria, such as the distribution of the random parameters (`randvars`), the number of random draws (`n_draws`), and the availability of alternatives for the choice situations (`avail`). We set the optimization method as `L-BFGS-B` as this is a robust routine that usually helps solve convergence issues.  Once the estimation routine is completed, the `summary` method can be used to display the estimation results.

In [6]:
varnames=['ASC_CAR', 'ASC_TRAIN', 'CO', 'TT']
model = MixedLogit()
res = model.fit(
    X=df[varnames],
    y=df['CHOICE'],
    varnames=varnames,
    alts=df['alt'],
    ids=df['custom_id'],
    avail=df['AV'],
    panels=df["ID"],
    randvars={'TT': 'n'},
    n_draws=1500,
)
model.summary()

2025-07-04 14:06:12,313 INFO jaxlogit.mixed_logit: Starting data preparation, including generation of 1500 random draws for each random variable and observation.


INFO:2025-07-04 14:06:12,397:jax._src.xla_bridge:752: Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
2025-07-04 14:06:12,397 INFO jax._src.xla_bridge: Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
2025-07-04 14:06:14,522 INFO jaxlogit.mixed_logit: Compiling log-likelihood function.
2025-07-04 14:06:15,121 INFO jaxlogit.mixed_logit: Compilation finished, init neg_loglike = 7309.54
2025-07-04 14:06:15,124 INFO jaxlogit._optimize: Running minimization with method L-BFGS-B
2025-07-04 14:06:15,886 INFO jaxlogit._optimize: Iter 1, fun = 5509.545, |grad| = 1054.174
2025-07-04 14:06:16,336 INFO jaxlogit._optimize: Iter 2, fun = 4969.103, |grad| = 571.378
2025-07-04 14:06:16,758 INFO jaxlogit._optimize: Iter 3, fun = 4611.336, |grad| = 275.850
2025-07-04 14:06:17,224 INFO jaxlogit._optimize: Iter 4, fun = 447

    Message: CONVERGENCE: RELATIVE REDUCTION OF F <= FACTR*EPSMCH
    Iterations: 16
    Function evaluations: 17
Estimation time= 18.7 seconds
---------------------------------------------------------------------------
Coefficient              Estimate      Std.Err.         z-val         P>|z|
---------------------------------------------------------------------------
ASC_CAR                 0.2831119     0.0560481     5.0512331       4.5e-07 ***
ASC_TRAIN              -0.5722801     0.0794782    -7.2004683      6.65e-13 ***
CO                     -1.6601727     0.0778872   -21.3151023      1.26e-97 ***
TT                     -3.2289922     0.1749822   -18.4532610      3.17e-74 ***
sd.TT                   3.6485578     0.1683478    21.6727368     9.28e-101 ***
---------------------------------------------------------------------------
Significance:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Log-Likelihood= -4359.218
AIC= 8728.436
BIC= 8762.536


The negative signs for the cost and time coefficients suggest that decision makers experience a general disutility with alternatives that have higher waiting times and costs, which conforms to the underlying decision making theory. Note that these estimates are highly consistent with those returned by Biogeme (https://biogeme.epfl.ch/examples/swissmetro/05normalMixtureIntegral.html)

## Electricity Dataset

The electricity dataset contains 4,308 choices among four electricity suppliers based on the attributes of the offered plans, which include prices(pf), contract lengths(cl), time of day rates (tod), seasonal rates(seas), as well as attributes of the suppliers, which include whether the supplier is local (loc) and well-known (wk). The data was collected through a survey where 12 different choice situations were presented to each participant. The multiple responses per participants were organized into panels. Given that some participants answered less than 12 of the choice situations, some panels are unbalanced, which `xlogit` is able to handle. [Revelt and Train (1999)](https://escholarship.org/content/qt1900p96t/qt1900p96t.pdf) provide a detailed description of this dataset. 

### Read data

The dataset is already in long format so no reshaping is necessary, it can be used directly in xlogit.

In [7]:
df = pd.read_csv("https://raw.github.com/arteagac/xlogit/master/examples/data/electricity_long.csv")
df

choice   id  alt  pf  cl  loc  wk  tod  seas  chid
0           0    1    1   7   5    0   1    0     0     1
1           0    1    2   9   1    1   0    0     0     1
2           0    1    3   0   0    0   0    0     1     1
3           1    1    4   0   5    0   1    1     0     1
4           0    1    1   7   0    0   1    0     0     2
...       ...  ...  ...  ..  ..  ...  ..  ...   ...   ...
17227       0  361    4   0   1    1   0    0     1  4307
17228       1  361    1   9   0    0   1    0     0  4308
17229       0  361    2   7   0    0   0    0     0  4308
17230       0  361    3   0   1    0   1    0     1  4308
17231       0  361    4   0   5    1   0    1     0  4308

[17232 rows x 10 columns]

### Fit the model

Note that the parameter `panels` was included in the `fit` function in order to take into account panel structure of this dataset during estimation.

In [8]:
varnames = ['pf', 'cl', 'loc', 'wk', 'tod', 'seas']
model = MixedLogit()
res = model.fit(
    X=df[varnames],
    y=df['choice'],
    varnames=varnames,
    ids=df['chid'],
    panels=df['id'],
    alts=df['alt'],
    n_draws=600,
    randvars={'pf': 'n', 'cl': 'n', 'loc': 'n', 'wk': 'n', 'tod': 'n', 'seas': 'n'}
)
model.summary()

2025-07-04 14:06:31,392 INFO jaxlogit.mixed_logit: Starting data preparation, including generation of 600 random draws for each random variable and observation.
2025-07-04 14:06:33,713 INFO jaxlogit.mixed_logit: Compiling log-likelihood function.
2025-07-04 14:06:34,210 INFO jaxlogit.mixed_logit: Compilation finished, init neg_loglike = 5620.08
2025-07-04 14:06:34,212 INFO jaxlogit._optimize: Running minimization with method L-BFGS-B
2025-07-04 14:06:35,047 INFO jaxlogit._optimize: Iter 1, fun = 5377.286, |grad| = 814.514
2025-07-04 14:06:35,499 INFO jaxlogit._optimize: Iter 2, fun = 5324.697, |grad| = 662.525
2025-07-04 14:06:36,028 INFO jaxlogit._optimize: Iter 3, fun = 5276.612, |grad| = 517.721
2025-07-04 14:06:36,376 INFO jaxlogit._optimize: Iter 4, fun = 5196.625, |grad| = 3295.426
2025-07-04 14:06:36,725 INFO jaxlogit._optimize: Iter 5, fun = 5036.259, |grad| = 921.954
2025-07-04 14:06:37,151 INFO jaxlogit._optimize: Iter 6, fun = 4937.728, |grad| = 971.137
2025-07-04 14:06:37,5

    Message: CONVERGENCE: RELATIVE REDUCTION OF F <= FACTR*EPSMCH
    Iterations: 74
    Function evaluations: 83
Estimation time= 42.0 seconds
---------------------------------------------------------------------------
Coefficient              Estimate      Std.Err.         z-val         P>|z|
---------------------------------------------------------------------------
pf                     -0.9971888     0.0378214   -26.3657437     3.73e-142 ***
cl                     -0.2196766     0.0255173    -8.6089275      1.02e-17 ***
loc                     2.2901968     0.1263079    18.1318519      7.26e-71 ***
wk                      1.6942744     0.0961503    17.6210989      3.66e-67 ***
tod                    -9.6751361     0.3350899   -28.8732511     9.81e-168 ***
seas                   -9.6960628     0.3246480   -29.8663842     2.66e-178 ***
sd.pf                   0.2207231     0.0191063    11.5523498      2.01e-30 ***
sd.cl                   0.4115696     0.0254622    16.1639602       

## Fishing Dataset

This example illustrates the estimation of a Mixed Logit model for choices of 1,182 individuals for sport fishing modes using `xlogit`. The goal is to analyze the market shares of four alternatives (i.e., beach, pier, boat, and charter) based on their cost and fish catch rate. [Cameron (2005)](http://cameron.econ.ucdavis.edu/mmabook/mma.html) provides additional details about this dataset. The following code illustrates how to use `xlogit` to estimate the model parameters.

### Read data

The data to be analyzed can be imported to Python using any preferred method. In this example, the data in CSV format was imported using the popular `pandas` Python package. However, it is worth highlighting that `xlogit` does not depend on the `pandas` package, as `xlogit` can take any array-like structure as input. This represents an additional advantage because `xlogit` can be used with any preferred dataframe library, and not only with `pandas`.

In [9]:
import pandas as pd
df = pd.read_csv("https://raw.github.com/arteagac/xlogit/master/examples/data/fishing_long.csv")
df

id      alt  choice      income    price   catch
0        1    beach       0  7083.33170  157.930  0.0678
1        1     boat       0  7083.33170  157.930  0.2601
2        1  charter       1  7083.33170  182.930  0.5391
3        1     pier       0  7083.33170  157.930  0.0503
4        2    beach       0  1249.99980   15.114  0.1049
...    ...      ...     ...         ...      ...     ...
4723  1181     pier       0   416.66668   36.636  0.4522
4724  1182    beach       0  6250.00130  339.890  0.2537
4725  1182     boat       1  6250.00130  235.436  0.6817
4726  1182  charter       0  6250.00130  260.436  2.3014
4727  1182     pier       0  6250.00130  339.890  0.1498

[4728 rows x 6 columns]

### Fit model

Once the data is in the `Python` environment, `xlogit` can be used to fit the model, as shown below. The `MultinomialLogit` class is imported from `xlogit`, and its constructor is used to initialize a new model. The `fit` method estimates the model using the input data and estimation criteria provided as arguments to the method's call. The arguments of the `fit` methods are described in [`xlogit`'s documentation](https://https://xlogit.readthedocs.io/en/latest/api/).


In [10]:
varnames = ['price', 'catch']
model = MixedLogit()
model.fit(
    X=df[varnames],
    y=df['choice'],
    varnames=varnames,
    alts=df['alt'],
    ids=df['id'],
    n_draws=1000,
    randvars={'price': 'n', 'catch': 'n'},
)
model.summary()

2025-07-04 14:07:13,596 INFO jaxlogit.mixed_logit: Starting data preparation, including generation of 1000 random draws for each random variable and observation.
2025-07-04 14:07:16,001 INFO jaxlogit.mixed_logit: Compiling log-likelihood function.
2025-07-04 14:07:16,332 INFO jaxlogit.mixed_logit: Compilation finished, init neg_loglike = 2586.70
2025-07-04 14:07:16,333 INFO jaxlogit._optimize: Running minimization with method L-BFGS-B
2025-07-04 14:07:16,491 INFO jaxlogit._optimize: Iter 1, fun = 2261.519, |grad| = 509.074
2025-07-04 14:07:16,579 INFO jaxlogit._optimize: Iter 2, fun = 2246.481, |grad| = 455.966
2025-07-04 14:07:16,673 INFO jaxlogit._optimize: Iter 3, fun = 2119.608, |grad| = 531.081
2025-07-04 14:07:16,813 INFO jaxlogit._optimize: Iter 4, fun = 2086.043, |grad| = 710.226
2025-07-04 14:07:16,946 INFO jaxlogit._optimize: Iter 5, fun = 2043.830, |grad| = 954.897
2025-07-04 14:07:17,186 INFO jaxlogit._optimize: Iter 6, fun = 2029.430, |grad| = 1228.491
2025-07-04 14:07:17,

    Message: CONVERGENCE: RELATIVE REDUCTION OF F <= FACTR*EPSMCH
    Iterations: 37
    Function evaluations: 54
Estimation time= 12.2 seconds
---------------------------------------------------------------------------
Coefficient              Estimate      Std.Err.         z-val         P>|z|
---------------------------------------------------------------------------
price                  -0.0272469     0.0022842   -11.9282999      4.71e-31 ***
catch                   1.3260449     0.1738921     7.6256762      4.97e-14 ***
sd.price                0.0102261     0.0021414     4.7753319      2.02e-06 ***
sd.catch                1.5661804     0.3755151     4.1707526      3.26e-05 ***
---------------------------------------------------------------------------
Significance:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Log-Likelihood= -1300.523
AIC= 2609.046
BIC= 2629.346


## Car Dataset

The fourth example uses a stated preference panel dataset for choice of car. Three alternatives are considered, with upto 6 choice situations per individual. This again is an unbalanced panel with responses of some individuals less than 6 situations. The dataset contains 8 explanaotry variables: price, operating cost, range, and binary indicators to indicate whether the car is electric, hybrid, and if performance is high or medium respectively. This dataset was taken from Kenneth Train's MATLAB codes for estimation of Mixed Logit models as shown in this link: https://eml.berkeley.edu/Software/abstracts/train1006mxlmsl.html

### Read data

In [11]:
import pandas as pd
import numpy as np

df = pd.read_csv("https://raw.github.com/arteagac/xlogit/master/examples/data/car100_long.csv")

Since price and operating cost need to be estimated with negative coefficients, we reverse the variable signs in the dataframe. 

In [12]:
df['price'] = -df['price']/10000
df['opcost'] = -df['opcost']
df

person_id  choice_id  alt  choice   price  opcost  range  ev  gas  \
0             1          1    1       0 -4.6763  -47.43    0.0   0    0   
1             1          1    2       1 -5.7209  -27.43    1.3   1    0   
2             1          1    3       0 -8.7960  -32.41    1.2   1    0   
3             1          2    1       1 -3.3768   -4.89    1.3   1    0   
4             1          2    2       0 -9.0336  -30.19    0.0   0    0   
...         ...        ...  ...     ...     ...     ...    ...  ..  ...   
4447        100       1483    2       0 -2.8036  -14.45    1.6   1    0   
4448        100       1483    3       0 -1.9360  -54.76    0.0   0    1   
4449        100       1484    1       1 -2.4054  -50.57    0.0   0    1   
4450        100       1484    2       0 -5.2795  -21.25    0.0   0    0   
4451        100       1484    3       0 -6.0705  -25.41    1.4   1    0   

      hybrid  hiperf  medhiperf  
0          1       0          0  
1          0       1          1  
2          0       0          1  
3          0       1          1  
4          1       0          1  
...      ...     ...        ...  
4447       0       0          0  
4448       0       1          1  
4449       0       0          0  
4450       1       0          1  
4451       0       0          0  

[4452 rows x 12 columns]

### Fit the model

In [ ]:
varnames = ['hiperf', 'medhiperf', 'price', 'opcost', 'range', 'ev', 'hybrid'] 
model = MixedLogit()
model.fit(
    X=df[varnames],
    y=df['choice'],
    varnames=varnames,
    alts=df['alt'],
    ids=df['choice_id'],
    panels=df['person_id'],
    randvars = {'price': 'ln', 'opcost': 'n',  'range': 'ln', 'ev':'n', 'hybrid': 'n'}, 
    n_draws = 1000,
)
model.summary()

2025-07-04 14:18:05,049 INFO jaxlogit.mixed_logit: Starting data preparation, including generation of 1000 random draws for each random variable and observation.
2025-07-04 14:18:06,082 INFO jaxlogit.mixed_logit: Compiling log-likelihood function.
2025-07-04 14:18:06,717 INFO jaxlogit.mixed_logit: Compilation finished, init neg_loglike = 1940.19
2025-07-04 14:18:06,719 INFO jaxlogit._optimize: Running minimization with method L-BFGS-B
2025-07-04 14:18:07,090 INFO jaxlogit._optimize: Iter 1, fun = 1831.796, |grad| = 490.910
2025-07-04 14:18:07,295 INFO jaxlogit._optimize: Iter 2, fun = 1675.649, |grad| = 276.954
2025-07-04 14:18:07,509 INFO jaxlogit._optimize: Iter 3, fun = 1566.488, |grad| = 183.438
2025-07-04 14:18:07,714 INFO jaxlogit._optimize: Iter 4, fun = 1505.994, |grad| = 189.795
2025-07-04 14:18:08,180 INFO jaxlogit._optimize: Iter 5, fun = 1408.467, |grad| = 406.549
2025-07-04 14:18:08,798 INFO jaxlogit._optimize: Iter 6, fun = 1384.970, |grad| = 546.910
2025-07-04 14:18:09,3

    Message: CONVERGENCE: RELATIVE REDUCTION OF F <= FACTR*EPSMCH
    Iterations: 131
    Function evaluations: 164
Estimation time= 45.9 seconds
---------------------------------------------------------------------------
Coefficient              Estimate      Std.Err.         z-val         P>|z|
---------------------------------------------------------------------------
hiperf                  0.1057423     0.0967948     1.0924373         0.275    
medhiperf               0.5713368     0.1004435     5.6881405      1.54e-08 ***
price                  -0.7405432     0.1406665    -5.2645315      1.61e-07 ***
opcost                  0.0119847     0.0057168     2.0964165        0.0362 *  
range                  -0.6711015     0.4033526    -1.6638087        0.0964 .  
ev                     -1.5934820     0.3357166    -4.7465089      2.27e-06 ***
hybrid                  0.7058535     0.1624154     4.3459763      1.48e-05 ***
sd.price                0.9461504     0.1202474     7.8683639     

## References

- Bierlaire, M. (2018). PandasBiogeme: a short introduction. EPFL (Transport and Mobility Laboratory, ENAC).

- Brathwaite, T., & Walker, J. L. (2018). Asymmetric, closed-form, finite-parameter models of multinomial choice. Journal of Choice Modelling, 29, 78–112. 

- Cameron, A. C., & Trivedi, P. K. (2005). Microeconometrics: methods and applications. Cambridge university press.

- Croissant, Y. (2020). Estimation of Random Utility Models in R: The mlogit Package. Journal of Statistical Software, 95(1), 1-41.

- Revelt, D., & Train, K. (1999). Customer-specific taste parameters and mixed logit. University of California, Berkeley.

